src/library_manager/llext_manager.c

```diff
@@ -108,130 +108,161 @@ static int llext_manager_load_module(uint32_t module_id, const struct sof_man_mo
+	size_t rodata_size = ctx->segment[LIB_MANAGER_RODATA].size;
+
+	/* Writable data (.data, .bss and others) */
+	void __sparse_cache *va_base_data = (void __sparse_cache *)
+		ctx->segment[LIB_MANAGER_DATA].addr;
+	void *src_data = (void *)(load_base +
+				  ctx->segment[LIB_MANAGER_DATA].file_offset);
+	size_t data_size = ctx->segment[LIB_MANAGER_DATA].size;
+
+	/* .bss, should be within writable data above */
+	void __sparse_cache *bss_addr = (void __sparse_cache *)
+		ctx->segment[LIB_MANAGER_BSS].addr;
+	size_t bss_size = ctx->segment[LIB_MANAGER_BSS].size;
 	int ret;
 
+	/* Check, that .bss is within .data */
+	if (bss_size &&
+	    ((uintptr_t)bss_addr + bss_size < (uintptr_t)va_base_data ||
+	     (uintptr_t)bss_addr >= (uintptr_t)va_base_data + data_size)) {
+		tr_err(&lib_manager_tr, ".bss %#x @ %p isn't within writable data %#x @ %p!",
+		       bss_size, bss_addr, data_size, (void *)va_base_data);
+		return -EPROTO;
+	}
+
 	/* Copy Code */
-	ret = llext_manager_load_data_from_storage(va_base_text, src_txt, st_text_size,
-						   SYS_MM_MEM_PERM_RW | SYS_MM_MEM_PERM_EXEC);
+	ret = llext_manager_load_data_from_storage(va_base_text, src_txt, text_size,
+						   SYS_MM_MEM_PERM_EXEC);
 	if (ret < 0)
 		return ret;
 
-	/* Copy RODATA */
+	/* Copy read-only data */
 	ret = llext_manager_load_data_from_storage(va_base_rodata, src_rodata,
-						   st_rodata_size, SYS_MM_MEM_PERM_RW);
+						   rodata_size, 0);
+	if (ret < 0)
+		goto e_text;
+
+	/* Copy writable data */
+	ret = llext_manager_load_data_from_storage(va_base_data, src_data,
+						   data_size, SYS_MM_MEM_PERM_RW);
 	if (ret < 0)
-		llext_manager_align_unmap(va_base_text, st_text_size);
+		goto e_rodata;
+
+	memset((__sparse_force void *)ctx->segment[LIB_MANAGER_BSS].addr, 0,
+	       ctx->segment[LIB_MANAGER_BSS].size);
+
+	return 0;
+
+e_rodata:
+	llext_manager_align_unmap(va_base_rodata, rodata_size);
+e_text:
+	llext_manager_align_unmap(va_base_text, text_size);
 
 	return ret;
 }
 
 static int llext_manager_unload_module(uint32_t module_id, const struct sof_man_module *mod)
 {
 	struct lib_manager_mod_ctx *ctx = lib_manager_get_mod_ctx(module_id);
+	/* Executable code (.text) */
 	void __sparse_cache *va_base_text = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
-	size_t st_text_size = ctx->segment_size[SOF_MAN_SEGMENT_TEXT];
+		ctx->segment[LIB_MANAGER_TEXT].addr;
+	size_t text_size = ctx->segment[LIB_MANAGER_TEXT].size;
+
+	/* Read-only data (.rodata, etc.) */
 	void __sparse_cache *va_base_rodata = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
-	size_t st_rodata_size = ctx->segment_size[SOF_MAN_SEGMENT_RODATA];
+		ctx->segment[LIB_MANAGER_RODATA].addr;
+	size_t rodata_size = ctx->segment[LIB_MANAGER_RODATA].size;
+
+	/* Writable data (.data, .bss, etc.) */
+	void __sparse_cache *va_base_data = (void __sparse_cache *)
+		ctx->segment[LIB_MANAGER_RODATA].addr;
```

Comment:
@lyakh Shouldn't this be LIB_MANAGER_DATA ? va_base_rodata already set above.

Reply:
@kv2019i oops, thanks for catching! The clean-up path - easy to miss without wide-scale automated testing...